src/_tkagg.cpp

```diff
@@ -232,6 +264,90 @@ static PyMethodDef functions[] = {
     { NULL, NULL } /* sentinel */
 };
 
+#ifdef DYNAMIC_TKINTER
+// Functions to fill global TCL / Tk function pointers from tkinter module.
+
+#include <dlfcn.h>
+
+#if PY3K
+#define TKINTER_PKG "tkinter"
+#define TKINTER_MOD "_tkinter"
+// From module __file__ attribute to char *string for dlopen.
+#define FNAME2CHAR(s) (PyBytes_AsString(PyUnicode_EncodeFSDefault(s)))
+#else
+#define TKINTER_PKG "Tkinter"
+#define TKINTER_MOD "tkinter"
+// From module __file__ attribute to char *string for dlopen
+#define FNAME2CHAR(s) (PyString_AsString(s))
+#endif
+
+void *_dfunc(void *lib_handle, const char *func_name)
+{
+    // Load function, unless there has been a previous error.  If so, then
+    // return NULL.  If there is an error loading the function, return NULL
+    // and set error flag.
+    static int have_error = 0;
+    void *func = NULL;
+    if (have_error == 0) {
+        // reset errors
+        dlerror();
+        func = dlsym(lib_handle, func_name);
+        const char *error = dlerror();
+        if (error != NULL) {
+            PyErr_SetString(PyExc_RuntimeError, error);
+            have_error = 1;
+        }
+    }
+    return func;
+}
+
+int _func_loader(void *tkinter_lib)
+{
+    // Fill global function pointers from dynamic lib.
+    // Return 0 fur success; 1 otherwise.
+    TCL_CREATE_COMMAND = (tcl_cc) _dfunc(tkinter_lib, "Tcl_CreateCommand");
+    TCL_APPEND_RESULT = (tcl_app_res) _dfunc(tkinter_lib, "Tcl_AppendResult");
+    TK_MAIN_WINDOW = (tk_mw) _dfunc(tkinter_lib, "Tk_MainWindow");
+    TK_FIND_PHOTO = (tk_fp) _dfunc(tkinter_lib, "Tk_FindPhoto");
+    TK_PHOTO_PUTBLOCK = (tk_ppb_nc) _dfunc(tkinter_lib, "Tk_PhotoPutBlock_NoComposite");
+    TK_PHOTO_BLANK = (tk_pb) _dfunc(tkinter_lib, "Tk_PhotoBlank");
+    return (TK_PHOTO_BLANK == NULL);
+}
+
+int load_tkinter_funcs(void)
+{
+    // Load tkinter global funcs from tkinter compiled module.
+    // Return 0 for success, non-zero for failure.
+    int ret = -1;
+    PyObject *pModule, *pSubmodule, *pString;
+
+    pModule = PyImport_ImportModule(TKINTER_PKG);
```

Comment:
It's a matter of personal preference, but I generally like to avoid the excessive nesting by:

1) setting all `PyObject *` to NULL in their declaration
2) doing:

```
if (pModule == NULL) {
  goto exit;
}
```

3) having

```
exit:
   Py_XDECREF(pString);
   Py_XDECREF(pSubmodule);
   Py_XDECREF(pModule);

   return ret;
```

at the bottom